internal/querynode/meta_replica.go

```diff
@@ -478,25 +479,40 @@ func (replica *metaReplica) getSegmentIDs(partitionID UniqueID, segType segmentT
 }
 
 // getSegmentIDsByVChannel returns segment ids which virtual channel is vChannel
-func (replica *metaReplica) getSegmentIDsByVChannel(partitionID UniqueID, vChannel Channel) ([]UniqueID, error) {
+// if partitionIDs is empty, it means that filtering by partitionIDs is not required.
+func (replica *metaReplica) getSegmentIDsByVChannel(partitionIDs []UniqueID, vChannel Channel, segType segmentType) ([]UniqueID, error) {
 	replica.mu.RLock()
 	defer replica.mu.RUnlock()
-	segmentIDs, err := replica.getSegmentIDsPrivate(partitionID, segmentTypeGrowing)
-	if err != nil {
-		return nil, err
+
+	var segments map[UniqueID]*Segment
+	var ret []UniqueID
+
+	filterPartition := len(partitionIDs) != 0
+	switch segType {
+	case segmentTypeGrowing:
+		segments = replica.growingSegments
+	case segmentTypeSealed:
+		segments = replica.sealedSegments
+	default:
+		return nil, fmt.Errorf("unexpected segment type, segmentType = %s", segType.String())
 	}
-	segmentIDsTmp := make([]UniqueID, 0)
-	for _, segmentID := range segmentIDs {
-		segment, err := replica.getSegmentByIDPrivate(segmentID, segmentTypeGrowing)
-		if err != nil {
-			return nil, err
-		}
+
+	for _, segment := range segments {
 		if segment.vChannelID == vChannel {
-			segmentIDsTmp = append(segmentIDsTmp, segment.ID())
+			if filterPartition {
+				partitionID := segment.partitionID
+				for _, partID := range partitionIDs {
+					if partitionID == partID {
+						ret = append(ret, segment.ID())
+						break
+					}
```

Comment:
you can construct a map to filter partition, do not traverse all partition ids per segment